File: app/src/main/java/com/aayush/filmipedia/util/datasource/factory/GenreDataFactory.java
package com.aayush.filmipedia.util.datasource.factory;

import com.aayush.filmipedia.FilmipediaApplication;
import com.aayush.filmipedia.model.Genre;
import com.aayush.filmipedia.util.datasource.GenreDataSource;

import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;
import io.reactivex.disposables.CompositeDisposable;

public class GenreDataFactory extends DataSource.Factory<Long, Genre> {
    private MutableLiveData<GenreDataSource> mutableLiveData;
    private GenreDataSource genreDataSource;
    private FilmipediaApplication application;
    private CompositeDisposable compositeDisposable;

    public GenreDataFactory(FilmipediaApplication application,
                              CompositeDisposable compositeDisposable) {
        this.application = application;
        this.compositeDisposable = compositeDisposable;
        mutableLiveData = new MutableLiveData<>();
    }

    @Override
    public DataSource<Long, Genre> create() {
        genreDataSource = new GenreDataSource(application, compositeDisposable);
        mutableLiveData.postValue(genreDataSource);
        return genreDataSource;
    }

    public MutableLiveData<GenreDataSource> getMutableLiveData() {
        return mutableLiveData;
    }
}

File: app/src/main/java/com/aayush/filmipedia/util/datasource/factory/GenreDataFactory.kt
package com.aayush.filmipedia.util.datasource.factory

import androidx.lifecycle.MutableLiveData
import androidx.paging.DataSource
import com.aayush.filmipedia.FilmipediaApplication
import com.aayush.filmipedia.model.Genre
import com.aayush.filmipedia.util.datasource.GenreDataSource
import io.reactivex.disposables.CompositeDisposable

class GenreDataFactory(
    private val application: FilmipediaApplication,
    private val compositeDisposable: CompositeDisposable
): DataSource.Factory<Long, Genre>() {
    val mutableLiveData = MutableLiveData<GenreDataSource>()
    private lateinit var genreDataSource: GenreDataSource

    override fun create(): DataSource<Long, Genre> {
        genreDataSource = GenreDataSource(application, compositeDisposable)
        mutableLiveData.postValue(genreDataSource)
        return genreDataSource
    }
}